lib/class-controller.php

```diff
@@ -25,229 +27,194 @@ class Controller {
 	}
 
 	/**
-	 * Get an instance of the class.
-	 *
-	 * @return Controller
+	 * Registers action and/or filter hooks with WordPress.
 	 */
-	public static function instance(): Controller {
-		if ( ! isset( self::$instance ) ) {
-			self::$instance = new self();
-		}
-		return self::$instance;
+	public function hook(): void {
+		add_action( 'init', [ $this, 'action__init' ], 1000 );
 	}
 
 	/**
-	 * Loads an instance of an Adapter into the controller.
+	 * Loads an adapter, either using the given adapter, or dynamically based
+	 * on environment settings.
 	 *
-	 * @param Adapter $adapter The adapter to load.
+	 * @param ?Adapter $adapter Optional. The adapter to load. Defaults to dynamic load.
 	 */
-	public function load_adapter( Adapter $adapter ): void {
-		$this->adapter = $adapter;
+	public function load_adapter( ?Adapter $adapter = null ): void {
+		if ( ! is_null( $adapter ) ) {
+			$this->adapter = $adapter;
+		} elseif ( defined( 'VIP_ENABLE_VIP_SEARCH' ) && VIP_ENABLE_VIP_SEARCH ) {
+			$this->adapter = Factory::vip_enterprise_search_adapter();
+		} else {
+			$this->adapter = Factory::generic_adapter();
+		}
 	}
 
 	/**
-	 * Map a given field to the Elasticsearch index.
+	 * Restricts searchable post types to the provided list.
+	 *
+	 * @param string[] $post_types The array of post types to restrict search to.
 	 *
-	 * @param  string $field The field to map.
-	 * @return string The mapped field.
+	 * @return Controller The instance of the class to allow for chaining.
 	 */
-	public function map_field( $field ) {
-		return $this->adapter->map_field( $field );
+	public function restrict_post_types( array $post_types ): Controller {
+		$this->adapter->restrict_post_types( $post_types );
+
+		return $this;
 	}
 
 	/**
-	 * Map a meta field. This will swap in the data type.
+	 * Restricts searchable taxonomies to the provided list.
 	 *
-	 * @param  string $meta_key Meta key to map.
-	 * @param  string $type Data type to map.
-	 * @return string The mapped field.
+	 * @param string[] $taxonomies The array of taxonomies to restrict search to.
+	 *
+	 * @return Controller The instance of the class to allow for chaining.
 	 */
-	public function map_meta_field( string $meta_key, string $type = '' ): string {
-		return $this->adapter->map_meta_field( $meta_key, $type );
+	public function restrict_taxonomies( array $taxonomies ): Controller {
+		$this->adapter->restrict_taxonomies( $taxonomies );
+
+		return $this;
 	}
 
 	/**
-	 * Map a taxonomy field. This will swap in the taxonomy name.
-	 *
-	 * @param  string $taxonomy Taxonomy to map.
-	 * @param  string $field Field to map.
-	 * @return string The mapped field.
+	 * Unregisters action and/or filter hooks with WordPress.
 	 */
-	public function map_tax_field( string $taxonomy, string $field ): string {
-		return $this->adapter->map_tax_field( $taxonomy, $field );
+	public function unhook(): void {
+		remove_action( 'init', [ $this, 'action__init' ], 99 );
```

Comment:
totally, I like the idea of offloading that work to a trait - I think the biggest case to make here is that missing remove action priorities is often a slip that takes a ton of time to troubleshoot so going the extra mile to ensure that we avoid those is worth the effort. Also, that would make the set up a little cleaner, instead of explicitly defining the hook/unhook methods, those could just be handled by the trait and the filters would only need to be set up in one place in the extending classes.